### PR TITLE
OS: Deprecate ConstString in favour of std::string

### PR DESCRIPTION
**Pull request by drdanz**
The `YARP_WRAP_STL_STRING` and `YARP_WRAP_STL_STRING_INLINE` are now always undefined, and they are now considered deprecated.

We need to test the impact of this patch on MSVC before merging, since this is supposed to cause several warnings. @pattacini @randaz81 can you test this?

### CODE REVIEW

**Comment by claudiofantacci** [comment]
> According to appveyor, this change introduces ~5300 C4251 warnings 😨
> I'm not sure how to handle this, should we just remove the warning? But there are cases when this warning actually makes sense? cc @traversaro

My understanding so far is that if we compile our DLL with `/MD[d]` flag (dynamic link to the Windows CRT), we can ignore the warning (but still, not 100% sure about it). We may need to discuss it f2f.

**Comment by traversaro** [comment]
I think we can just suppress the warning. Exposing methods with STL data structures can be problematic when compiling with a static CRT or combining libraries linking different CRTs, but I suspect we already do not support this, we just need to be clear about this in the documentation. 

A much tricker thing to manager is **how** to suppress the warning, in a way that propagates correctly to users of the library, without suppressing the warning if it is introduced in the user source code. 
I am afraid the only suitable way of doing so is to use [`#pragma warning`](https://msdn.microsoft.com/en-us/library/2c8f766e.aspx) but that basically means modifying **all headers** that generate that warning, and I am not sure how feasible this is.

**Comment by traversaro** [comment]
Obligatory link to a great recap slide on this : https://docs.google.com/presentation/d/14Ags6mV-xQbN6OIfVOPcvDOW9M2WlgwGls-QYhZe2GQ/edit?usp=sharing . 

/hattip @mxgrey (You should set up a website just containing this slide).

**Comment by mxgrey** [comment]
To suppress the warning in a good, portable, transferable way, I recommend defining a pair of macros, something like

```
#ifdef _MSC_VER
#define YARP_SUPPRESS_INTERFACE_WARNING // TODO: add code to suppress C4251 for MSVC
#define YARP_CONTINUE_INTERFACE_WARNING // TODO: add code to pop the warning stack for MSVC
#else // This will be used for Clang, GCC, and other non-MSVC compilers
#define YARP_SUPPRESS_INTERFACE_WARNING // leave this blank
#define YARP_CONTINUE_INTERFACE_WARNING // leave this blank
#endif
```

Then you can wrap those two macros directly around the STL class members that trigger the warning. This can still mean modifying a *lot* of headers depending on how many places you're using STL class members, but at least the modifications will be reasonably clean.

**Comment by traversaro** [comment]
Thanks @mxgrey for chiming in. I like both forms  of `YARP_SUPPRESS_INTERFACE_WARNING`, even if they will require a lot of headers modifications. I am not a big fan of `target_compile_options(<target> PUBLIC /wd4251)`, as it will suppress the warning also for the all downstream code contained in the library/executable that will link YARP.

**Comment by claudiofantacci** [comment]
> This needs to be tested, though, any MSVC user wants to test it?

I can test this. We can plani it f2f 👍

**Comment by drdanz** [comment]
As discussed, I'm changing this in order to deprecate `ConstString` in favour of `std::string`

File: bindings/yarp.i [hunk 1]
```diff
@@ -42,10 +42,10 @@
   %include "std_vector.i"
 #endif
 
-// Try to make yarp::os::ConstString act like std::string
+// Try to make std::string act like std::string
```
Comment (traversaro) [inline bindings/yarp.i]: This is weird now. :D

Comment (drdanz) [reply bindings/yarp.i]: :rofl:
I'm not sure if this part should be removed from the bindings...

**Comment by drdanz** [comment]
After some testing, this is the version I like the most:

```c++
#if defined(_MSC_VER)
# define YARP_SUPPRESS_INTERFACE_WARNING(...) \
     __pragma(warning(suppress:4251)) \
     ##__VA_ARGS__
#else
# define YARP_SUPPRESS_INTERFACE_WARNING(...) ##__VA_ARGS__
#endif
```

Then, all the declarations causing the warning are wrapped in this way:

```c++
    YARP_SUPPRESS_INTERFACE_WARNING(std::string foo);
```

@traversaro what do you think?

**Comment by drdanz** [comment]
or in alternative:

```c++
    YARP_SUPPRESS_INTERFACE_WARNING(std::string) foo;
```

I think this is my favourite version

**Comment by claudiofantacci** [comment]
> or in alternative:
> 
>     YARP_SUPPRESS_INTERFACE_WARNING(std::string) foo;
> 
> I think this is my favourite version

IMHO, this version is much clearer.

**Comment by drdanz** [comment]
> IMHO, this version is much clearer.

Unfortunately I was wrong, they don't work

**Comment by drdanz** [comment]
Actually, I was wrong about being wrong, it works... :rofl:

**Comment by traversaro** [comment]
~~~
YARP_SUPPRESS_INTERFACE_WARNING(std::string) foo;
~~~
If this  works this version is quite cool, much better than everything else.

**Comment by drdanz** [comment]
It works, but unfortunately this doesn't:

```c++
static const YARP_SUPPRESS_DLL_INTERFACE_WARNING(std::string) foo;
```

instead I have to do

```c++
YARP_SUPPRESS_DLL_INTERFACE_WARNING(static const std::string) foo;
```

That isn't really exciting and makes this option cleaner:

```c++
YARP_SUPPRESS_DLL_INTERFACE_WARNING(static const std::string foo);
```

**Comment by drdanz** [comment]
Also I found out that this version suppress the warning also for the following line, but to be honest I couldn't care less... :laughing:

**Comment by drdanz** [comment]
```c++
#if defined(_MSC_VER)
# define YARP_DISABLE_DLL_INTERFACE_WARNING __pragma(warning(disable:4251))
# define YARP_SUPPRESS_DLL_INTERFACE_WARNING __pragma(warning(suppress:4251))
#else
# define YARP_DISABLE_DLL_INTERFACE_WARNING
# define YARP_SUPPRESS_DLL_INTERFACE_WARNING
#endif

#define YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(x) YARP_SUPPRESS_DLL_INTERFACE_WARNING \
                                                   x
```

Used as

```cpp
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string) foo;
```
```cpp
YARP_SUPPRESS_DLL_INTERFACE_WARNING
    static const std::string bar;
```
```cpp
YARP_WARNING_PUSH
YARP_DISABLE_DLL_INTERFACE_WARNING
    std::string baz;
    std::string zot;
YARP_WARNING_POP
```

I still have a doubt about using `x` or `...` + `##VA_ARGS` for the `YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG` macro.
The second option allows to do uglier things, for example these are all valid

```cpp
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string foo);
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string foo;)
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(static const std::string foo);
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(static const std::string /*hello world*/ foo);
```

But on the other hand, this cannot work with the first option:

```cpp
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::map<int, int>) foo;
```

due to the `,` in the template.

Options:

1) Add a `typedef` for the type

```c++

    typedef std::map<int, int> fooMap
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(fooMap) foo;
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string) bar;
```

2) Add a `YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARGS` macro to use in these cases

```c++
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARGS(std::map<int, int>) foo;
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string) bar;
```

3) Use `YARP_SUPPRESS_DLL_INTERFACE_WARNING`

```c++
    YARP_SUPPRESS_DLL_INTERFACE_WARNING
    std::map<int, int> foo;
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string) bar;
```

4) Use `...` in `YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG`

```c++
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::map<int, int>) foo;
    YARP_SUPPRESS_DLL_INTERFACE_WARNING_ARG(std::string) bar;
```

Option number 2 is my favourite, @traversaro, @claudiofantacci what do you think?

**Comment by traversaro** [comment]
:+1: for 2 .

**Pull request merged**